File: c/hash-table/minimum-window-substring.c
char* minWindow(char* s, char* t) {
    int i, start = 0;
    int s_len = strlen(s);
    int t_len = strlen(t);
    int min_index, min_window = s_len + 1;
    int cur_match = 0;
    char *ret = "";
    int t_map[256] = {0};
    int cur_map[256] = {0};

    for(i = 0; i < t_len; ++i)
        t_map[t[i]]++;

    for(i = 0; i < s_len; ++i){
        if(t_map[s[i]] == 0)
            continue;

        cur_map[s[i]]++;
        if(cur_map[s[i]] <= t_map[s[i]])
            cur_match++;

        if(cur_match != t_len)
            continue;

        while(!t_map[s[start]] ||
            cur_map[s[start]] > t_map[s[start]]){
            if(t_map[s[start]])
                cur_map[s[start]]--;
            start++;
        }

        if(i - start + 1 < min_window){
            min_window = i - start + 1;
            min_index = start;
        }
        cur_map[s[start]]--;
        start++;
        cur_match--;
    }

    if(min_window != s_len + 1){
        ret = malloc(sizeof(char) * (min_window + 1));
        memcpy(ret, s + min_index, min_window);
        ret[min_window] = 0;
    }

    return ret;
}

File: python/hash-table/minimum-window-substring.py
class Solution(object):
    def minWindow(self, s, t):
        """
        :type s: str
        :type t: str
        :rtype: str
        """
        init_hash = {}
        cur_hash = dict.fromkeys(t,0)
        for ch in t:
            init_hash[ch] = init_hash.get(ch, 0) + 1
        end = 0
        start = 0
        t_match = 0
        ret = ""
        for end in range(len(s)):
            if init_hash.get(s[end], 0) == 0:
                continue
            cur_hash[s[end]] += 1
            if cur_hash[s[end]] <= init_hash[s[end]]:
                t_match += 1
            if t_match < len(t):
                continue
            while init_hash.get(s[start], 0) == 0 or \
                cur_hash[s[start]] > init_hash[s[start]]:
                if cur_hash.get(s[start], 0) > 0:
                    cur_hash[s[start]] -= 1
                start += 1
            if len(ret) == 0 or end - start + 1 < len(ret):
                ret = s[start:end+1]
            t_match -= 1
            cur_hash[s[start]] -= 1
            start += 1

        return ret
